JDK-8281023: NMT integration into pp debug command does not work

tstuefe opened the pull request:
JDK-8280289 enhanced the debug pp() command to use NMT if enabled, and to print NMT related info. That is useful, but there are some issues.

On debug, it just asserts, since the empty reserved region we create to hold the output of the mmap-search is created with address=NULL:

```
(gdb) call pp(0x7ffff010b030)

"Executing pp"

Thread 2 "java" received signal SIGSEGV, Segmentation fault.
0x00007ffff6721a71 in VirtualMemoryRegion::VirtualMemoryRegion (this=this@entry=0x7ffff5bb2620, addr=addr@entry=0x0, size=size@entry=0) at /shared/projects/openjdk/jdk-jdk/source/src/hotspot/share/services/virtualMemoryTracker.hpp:180
180 assert(addr != NULL, "Invalid address");
```

On release we don't assert and get further, but the use of SafeFetch is slightly wrong. It will deny us any NMT data about p if *p==0:

```
if (CanUseSafeFetchN() && SafeFetchN((intptr_t*)p, 0) != 0) {
```

This patch:
- fixes uses of SafeFetch
- changes the mmap-region-search-code to not require an empty ReservedMemoryRegion in order to avoid triggering the assert in virtualMemoryTracker.hpp:180
- adds a comment about the safe use of pp() in gdb (one needs to switch off signal handling of SIGSEGV for this to work)

Tests:
- I tested manually that pp works with different levels of NMT (Linux x64)
- GHAs in process

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8281023](https://bugs.openjdk.java.net/browse/JDK-8281023): NMT integration into pp debug command does not work


### Reviewers
 * [Zhengyu Gu](https://openjdk.java.net/census#zgu) (@zhengyu123 - **Reviewer**)
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7297/head:pull/7297` \
`$ git checkout pull/7297`

Update a local copy of the PR: \
`$ git checkout pull/7297` \
`$ git pull https://git.openjdk.java.net/jdk pull/7297/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7297`

View PR using the GUI difftool: \
`$ git pr show -t 7297`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7297.diff">https://git.openjdk.java.net/jdk/pull/7297.diff</a>

</details>

zhengyu123 commented on src/hotspot/share/services/virtualMemoryTracker.cpp:
```diff
@@ -674,26 +674,26 @@ bool VirtualMemoryTracker::walk_virtual_memory(VirtualMemoryWalker* walker) {
   walk_virtual_memory(&walker);
-  return walker.found_region();
+  return walker.region();
 }
```
Snapshot the region is for avoiding race pointed out by Ioi in code review, because other thread might release the region after walk.

zhengyu123 commented on src/hotspot/share/utilities/debug.cpp:
```diff
@@ -485,23 +485,29 @@ extern "C" JNIEXPORT void pp(void* p) {
+      //  SafeFetch test to check if reading is safe. This will generate a signal if it isn't. That signal normally
+      //  is handled quietly by the VM, but it will trip up the debugger. gdb will catch the signal and disable
+      //  the pp() command for further use.
+      // In order to avoid that, before invoking pp(), switch off SIGSEGV handling with "handle SIGSEGV nostop".
```
Ah, my .gdbinit has `handle SIGSEGV nostop noprint pass`

tstuefe commented:
@zhengyu123 thanks for looking at this.

I rewrote the patch and made printing inline within lock protection. For symmetry, I also moved the printing code on the malloc side to the malloc tracker. Added comments.

Note that I think @iklam was overcautious in the original discussion. We already risk signals in the debugging session by using SafeFetch to read the malloc header. Using a reference to a potentially dead ReservedMemoryRegion runs an additional - very low - risk of signals. So, if we really want to be cautious, we should not print out NMT information at all in pp. But I think NMT is very useful and worth the risk. All we risk is a slightly miffed debugger.

zhengyu123 commented on src/hotspot/share/services/mallocTracker.cpp:
```diff
@@ -289,3 +290,32 @@ void* MallocTracker::record_free(void* memblock) {
+// signals popping up, e.g. when writing an hs_err file.
+bool MallocTracker::print_pointer_information(const void* p, outputStream* st) {
+  assert(MemTracker::enabled(), "NMT must be enabled");
+  if (CanUseSafeFetchN() && os::is_readable_pointer(p)) {
```
`os::is_readable_pointer()` uses `CanUseSafeFetch32()`, you may want to check `CanUseSafeFetch32()` instead of `CanUseSafeFetchN()`.

tstuefe replied:
Good point. Done.

iklam commented:
> Note that I think @iklam was overcautious in the original discussion. We already risk signals in the debugging session by using SafeFetch to read the malloc header. Using a reference to a potentially dead ReservedMemoryRegion runs an additional - very low - risk of signals. So, if we really want to be cautious, we should not print out NMT information at all in pp. But I think NMT is very useful and worth the risk. All we risk is a slightly miffed debugger.

I wasn't trying to make life unnecessarily difficult, and I agree that printing out the NMT info is a great idea.

All I was suggesting was -- if there's a less risky way to do the printing and that's no overly complicated, we should do it that way. And the new code looks good to me.

tstuefe commented:
> > Note that I think @iklam was overcautious in the original discussion. We already risk signals in the debugging session by using SafeFetch to read the malloc header. Using a reference to a potentially dead ReservedMemoryRegion runs an additional - very low - risk of signals. So, if we really want to be cautious, we should not print out NMT information at all in pp. But I think NMT is very useful and worth the risk. All we risk is a slightly miffed debugger.
> 
> I wasn't trying to make life unnecessarily difficult, and I agree that printing out the NMT info is a great idea.
> 
> All I was suggesting was -- if there's a less risky way to do the printing and that's no overly complicated, we should do it that way. And the new code looks good to me.

Thank you for the review!

If pp() were used outside of debugging, I'd agree with your original estimate.

tstuefe commented:
Thanks @zhengyu123 and @iklam.

/integrate
